### PR TITLE
ProcessArrayCmp() should handle ScalarArray of constants and non constants

### PR DESCRIPTION
**Pull request by sambitesh**
When there is a comparison between a scalar ident and an array consisting of
constants and non constants, ORCA would not estimate the cardinality correctly.
This is because the ProcessArrayCmp() function would simply return the value of
CStatsPredUnsupported() when the array was not a ConstArray. This patch fixes
the issue by considering Arrays in which constants and non constants are
present together and makes sure that the cardinality is estimated correctly.

e.g.
```
CREATE TABLE bar (a int);

INSERT INTO bar SELECT i FROM generate_series(1,10000) i;
INSERT INTO bar SELECT i%10 FROM generate_series(1,10000) i;
ANALYZE bar;
```

The MCVs are in the range 0-9.

But below queries with and without MCV both estimated the same number of rows before this fix.

```sql
explain select * from bar where a in (1, (select count(*) from bar));
                                              QUERY PLAN
------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice3; segments: 3)  (cost=0.00..1407777.04 rows=8000 width=4)
   ->  Nested Loop  (cost=0.00..1407765.17 rows=2667 width=4)
         Join Filter: public.bar.a = ANY (ARRAY[1::bigint, (count((count())))])
         ->  Broadcast Motion 1:3  (slice2)  (cost=0.00..444.64 rows=3 width=8)
               ->  Aggregate  (cost=0.00..444.64 rows=1 width=8)
                     ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..444.64 rows=1 width=8)
                           ->  Aggregate  (cost=0.00..444.64 rows=1 width=8)
                                 ->  Table Scan on bar  (cost=0.00..443.40 rows=20000 width=1)
         ->  Table Scan on bar  (cost=0.00..443.40 rows=20000 width=4)
 Optimizer status: PQO version 2.51.3
(10 rows)
```

Plan produced for IN clause with non MCV and non constant value in IN clause
```sql
explain select * from bar where a in (1000, (select count(*) from bar));
                                              QUERY PLAN
------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice3; segments: 3)  (cost=0.00..1407777.04 rows=8000 width=4)
   ->  Nested Loop  (cost=0.00..1407765.17 rows=2667 width=4)
         Join Filter: public.bar.a = ANY (ARRAY[1000::bigint, (count((count())))])
         ->  Broadcast Motion 1:3  (slice2)  (cost=0.00..444.64 rows=3 width=8)
               ->  Aggregate  (cost=0.00..444.64 rows=1 width=8)
                     ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..444.64 rows=1 width=8)
                           ->  Aggregate  (cost=0.00..444.64 rows=1 width=8)
                                 ->  Table Scan on bar  (cost=0.00..443.40 rows=20000 width=1)
         ->  Table Scan on bar  (cost=0.00..443.40 rows=20000 width=4)
 Optimizer status: PQO version 2.51.3
(10 rows)
```

Signed-off-by: Sambitesh Dash <sdash@pivotal.io>

### CODE REVIEW

**Comment by sambitesh** [comment]
We will bump the ORCA version before merging.

File: libnaucrates/src/statistics/CStatsPredUtils.cpp [hunk 1]
```diff
@@ -918,7 +918,9 @@ CStatsPredUtils::ProcessArrayCmp
 	}
 	CExpression *pexprArray = CUtils::PexprScalarArrayChild(pexprPred);
 
-	BOOL fCompareToConst = CPredicateUtils::FCompareIdentToConstArray(pexprPred) || CPredicateUtils::FCompareCastIdentToConstArray(pexprPred);
+	BOOL fCompareToConst = CPredicateUtils::FCompareIdentToConstArray(pexprPred) ||
+						   CPredicateUtils::FCompareCastIdentToConstArray(pexprPred) ||
```
Comment (dhanashreek89) [inline libnaucrates/src/statistics/CStatsPredUtils.cpp]: Do we need `FCompareIdentToConstArray` now? Wouldn't `FCompareScalarIdentToConstAndScalarIdentArray()` cover it as well?

File: libgpopt/src/base/CCastUtils.cpp [hunk 2]
```diff
@@ -161,6 +161,18 @@ CCastUtils::FBinaryCoercibleCast
 	CScalarCast::PopConvert(pop)->FBinaryCoercible();
 }
 
+BOOL
+CCastUtils::FCast
```
Comment (hsyuan) [inline libgpopt/src/base/CCastUtils.cpp]: FCast -> FScalarCast

File: libgpopt/src/base/CCastUtils.cpp [hunk 3]
```diff
@@ -161,6 +161,18 @@ CCastUtils::FBinaryCoercibleCast
...
+	(
+	CExpression *pexpr
+	)
+{
+	GPOS_ASSERT(NULL != pexpr);
+	COperator *pop = pexpr->Pop();
+	
+	return COperator::EopScalarCast == pop->Eopid();
+}
+
```
Comment (dhanashreek89) [inline libgpopt/src/base/CCastUtils.cpp]: Why not use `CCastUtils::FCast()` ?

File: libgpopt/src/base/CCastUtils.cpp (hunk 2 above)
Comment (hsyuan) [reply libgpopt/src/base/CCastUtils.cpp]: In general, this function should be in CUtils.cpp

File: libnaucrates/src/statistics/CStatsPredUtils.cpp (hunk 1 above)
Comment (dhanashreek89) [inline libnaucrates/src/statistics/CStatsPredUtils.cpp]: variable name `fCompareToConst`  should be changed now that we support comparing array of constants and non-constants

File: libgpopt/src/base/CCastUtils.cpp [hunk 4]
```diff
@@ -161,6 +161,18 @@ CCastUtils::FBinaryCoercibleCast
 	CScalarCast::PopConvert(pop)->FBinaryCoercible();
```
Comment (dhanashreek89) [inline libgpopt/src/base/CCastUtils.cpp]: Could we rewrite `FBinaryCoercibleCast()` as follows now:
```
return FCast(pexpr) && CScalarCast::PopConvert(pop)->FBinaryCoercible();
```

File: libgpopt/src/base/CCastUtils.cpp (hunk 2 above)
Comment (dhanashreek89) [reply libgpopt/src/base/CCastUtils.cpp]: We moved all cast related utilities to from CUtils.cpp to CCastUtils.cpp in https://github.com/greenplum-db/gporca/pull/208

File: libnaucrates/src/statistics/CStatsPredUtils.cpp (hunk 1 above)
Comment (oarap) [reply libnaucrates/src/statistics/CStatsPredUtils.cpp]: True, `FCompareScalarIdentToConstAndScalarIdentArray` should do the job

**Comment by sambitesh** [comment]
@dhanashreek89 @hsyuan @oarap Addressed PR comments.

File: server/src/unittest/gpopt/minidump/CICGTest.cpp [hunk 5]
```diff
@@ -40,6 +40,8 @@ ULONG CICGTest::m_ulNegativeIndexApplyTestCounter = 0;
 // minidump files
 const CHAR *rgszFileNames[] =
 	{
+		"../data/dxl/minidump/CastedInClauseWithMCV.mdp",
+		"../data/dxl/minidump/InClauseWithMCV.mdp",
```
Comment (hsyuan) [inline server/src/unittest/gpopt/minidump/CICGTest.cpp]: please don't add tests into CICGTest

Comment (sambitesh) [reply server/src/unittest/gpopt/minidump/CICGTest.cpp]: We were thinking the same thing too. Any suggestions on where to move it?

**Comment by hsyuan** [comment]
You can create new test for testing arrays

~ ~ ~
Haisheng Yuan

**Pull request merged**